Matcher.js

```diff
@@ -98,44 +107,158 @@ function Matcher() {
+        /*
+         * This portion of the function checks for students who are "eligible" for being given filler workshops, in other words,
+         * Students who still have two available workshop slots (one for the filler workshop, and one for their highest preference)
+         */
+        var eligibleStudents = [];
         for (var i = 0; i < this.allStudents.length; i++) {
-            var thisStudent = this.allStudents[i];
-            for (var j = 0; j < this.sessionsPerWorkshop; j++) {
-                thisStudent.assignWorkshopSession(
-                    thisStudent.preferences[j],
-                    j
-                );
+            var tempStudent = this.allStudents[i];
+            if (tempStudent.numberAssigned() < 2 && !tempStudent.isAssigned(workshop)) {
+                eligibleStudents.push(tempStudent);
             }
         }
-    };
+
+        /*
+         * This portion of the function iterates through preferences 1 through 6 and assigns random students who have that preference
+         * open to the workshop until it reaches its quorum. If the preference for that student is full, then they are temporarily
+         * removed from the list of "eligible" students until the next preference iteration.
+         */
+        for (var j = 0; j < this.numberOfPreferences; j++) { // first preference first, etc
+            var currentStage = eligibleStudents.slice();
+            while (currentStage.length > 0) {
+                var randomIndex = Math.floor(Math.random() * currentStage.length)
+                var randomStudent = currentStage[randomIndex];
+                var currentPreference = randomStudent.preferences[j];
+                if (currentPreference === null) {
+                    currentStage.splice(randomIndex, 1);
+                    continue;
+                }
+                if (randomStudent.isAssigned(currentPreference)) {
+                    randomStudent.assignWorkshop(workshop);
+                }
+                else if (!currentPreference.isFull()) {
+                    randomStudent.assignWorkshop(workshop);
+                    randomStudent.assignWorkshop(currentPreference);
+                }
+                currentStage.splice(randomIndex, 1);
+                if (workshop.hasReachedQuorum()) {
+                    return;
+                }
+            }
+        }
+        Logger.log("Unable to fill workshop " + workshop.name);
+        Logger.log("Slots filled: " + workshop.slotsFilled + "/" + workshop.minimumFill);
+        Logger.log("Base Popularity: " + workshop.popularityScore);
+        throw new Error("Was unable to fill workshop " + workshop.name); // will handle this differently if it happens, EXTREMELY unlikely
+    }
+
+    /**
+     * The main matching algorithm, fills every workshop to its minimum with the students who prefer that workshop the most
+     */
+    this.reachMinimumForAll = function() {
+        this.fullEnough = [];
```

Comment:
We've mostly been using "Quorum" in the code to refer to workshops being "full enough", we may want to remain consistent with these variable names.